ddtrace/tracer/example_test.go

```diff
@@ -3,37 +3,49 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016 Datadog, Inc.
 
-package tracer
+package tracer_test
 
 import (
-	"io/ioutil"
-	"log"
+	"context"
+	"fmt"
+	"time"
 
 	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
 )
 
 // A basic example demonstrating how to start the tracer, as well as how
 // to create a root span and a child span that is a descendant of it.
 func Example() {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
 	// Start the tracer and defer the Stop method.
-	Start(WithAgentAddr("host:port"))
-	defer Stop()
+	tracer.Start()
+	defer tracer.Stop()
 
 	// Start a root span.
-	span := StartSpan("get.data")
+	span, ctx := tracer.StartSpanFromContext(ctx, "your.work")
 	defer span.Finish()
 
-	// Create a child of it, computing the time needed to read a file.
-	child := StartSpan("read.file", ChildOf(span.Context()))
-	child.SetTag(ext.ResourceName, "test.json")
+	// Create a child, using the context of the parent span.
+	child, ctx := tracer.StartSpanFromContext(ctx, "this.service")
+	child.SetTag(ext.ResourceName, "alarm")
 
-	// Perform an operation.
-	_, err := ioutil.ReadFile("~/test.json")
+	// Run some code.
+	err := yourCode(ctx)
 
-	// We may finish the child span using the returned error. If it's
-	// nil, it will be disregarded.
-	child.Finish(WithError(err))
-	if err != nil {
-		log.Fatal(err)
+	// Finish the span.
+	child.Finish(tracer.WithError(err))
```

Comment:
We usually strongly advise against finishing spans like this. We recommend that _always_ the finish call be paired with the span start call in a defer, to make sure that whatever happens in between the two, doesn't cause the span to remain unfinished, which can cause traces not to be sent, and maybe even memory issues.

If having an error complicates things, the ugly way out is to make the error a variable before the `defer` and finish as such:
```go
defer func(err error) {
    child.Finish(tracer.WithError(err))
}(err)
````

It's not pretty, I know, but it's safer.

Reply:
Also, this is a bit weird. The fact that the parent span and child span both finish simultaneously raised a question. I think your intention was to instrument `yourCode`. In that case, I think that everything `child` related should be inside the `yourCode` function. Then, you can also call it `span`. It's a bit unusual to call a span `child`.

Reply:
Great, thanks for letting me know 👍 
I'd like the example code to be as close to how we expect people to use our API in practice. I updated it, so LMK if there's anything in this code that still seems off, or not how we would expect the API to be used and I'll change it.

Reply:
I'm not sure this updated example does quite what we want? Since the `err` is evaluated at the time of defer, if `err` is changed later the value won't make it inside the child.Finish (example: https://go.dev/play/p/MMw9auPtNvJ)

I think it might be better to just capture `err` in the scope of the func like this: https://go.dev/play/p/G4sITU_VfOq

Reply:
Is there an issue with the current implementation in the most recent commit? I believe it's doing what you suggested, but lmk if I'm misunderstanding you.

Reply:
<img width="1370" alt="Screen Shot 2022-07-25 at 3 23 42 PM" src="https://user-images.githubusercontent.com/28789760/180858128-4050ec0e-c544-44b6-b46d-db0d585cc5d3.png">

If I set the context to end at 1 millisecond, instead of 1 second, I can see an error in the child span. I don't know if we should expect the parent span to also fail in a situation like this, but it at least shows how to set a child to have an error. I'm not 100% sure how we expect our API to be used in the case of an error, so please let me know if I'm making the wrong assumption here.

Reply:
👍 ah that's what I get for only looking at the comment :P Yeah the code itself looks fine to me! I'm also not sure on how exactly to expect it to behave, but it seems normal that an error in a span wouldn't bubble up?